msal/src/Microsoft.Identity.Client/CacheV2/TokenCacheV2.cs

```diff
@@ -0,0 +1,234 @@
+                    return output;
+                }
+                else
+                {
+                    // call into _storageManager and get the data to serialize...
+                    return _storageManager.Serialize();
+                }
+            }
+        }
+
+        public void Deserialize(byte[] unifiedState)
+        {
+            lock (_lock)
+            {
+                if (_storageManager == null)
+                {
+                    _deserializedCacheDataHolding.UnifiedState = new byte[unifiedState.Length];
+                    Array.Copy(unifiedState, _deserializedCacheDataHolding.UnifiedState, unifiedState.Length);
+                }
+                else
+                {
+                    // call into _storageManager and push the contents...
+                    _storageManager.Deserialize(unifiedState);
+                }
+            }
+        }
+
+        private ILegacyCachePersistence LegacyCachePersistence => _storageManager?.AdalLegacyCacheManager?.LegacyCachePersistence;
+
+        public CacheData SerializeUnifiedAndAdalCache()
+        {
+            lock (_lock)
+            {
+                if (_storageManager == null && LegacyCachePersistence == null)
+                {
+                    return _deserializedCacheDataHolding;
+                }
+
+                var cacheData = new CacheData();
+                if (_storageManager != null)
+                {
+                    cacheData.UnifiedState = _storageManager.Serialize();
+                }
+
+                if (LegacyCachePersistence != null)
+                {
+                    cacheData.AdalV3State = LegacyCachePersistence.LoadCache();
+                }
+
+                return cacheData;
+            }
+        }
+
+        public void DeserializeUnifiedAndAdalCache(CacheData cacheData)
+        {
+            lock (_lock)
+            {
+                if (_storageManager == null && LegacyCachePersistence == null)
+                {
+                    _deserializedCacheDataHolding = cacheData;
+                }
+
+                _storageManager?.Deserialize(cacheData.UnifiedState);
+                LegacyCachePersistence?.WriteCache(cacheData.AdalV3State);
+            }
+        }
+
+        /// <summary>
+        ///     This is called when the TokenCache is set into the clientapplicationbase and we need to
+        ///     hook it up since the storage manager is the actual source of truth for the cache...
+        /// </summary>
+        /// <param name="storageManager"></param>
+        internal void BindToStorageManager(IStorageManager storageManager)
+        {
+            if (storageManager == null)
+            {
+                throw new ArgumentNullException(nameof(storageManager));
+            }
+
+            lock (_lock)
+            {
+                _storageManager = storageManager;
+
+                // If the user has deserialized any data into the cache before binding...
```

Comment:
>If the user has deserialized any data into the cache before binding... [](start = 19, length = 70)

per above. not sure why this would not always be bound, probably my real question

Reply:
Because in our current model, the TokenCache is constructed by the developer using our libraries, so when it's constructed none of this is bound up.  Eventually TokenCacheV2 will replace the public TokenCache object for public consumption and we need to handle them constructing it and then handing it to us (where we then bind up the types).